src/primaite/agents/llm/agent.py

```diff
@@ -45,67 +52,182 @@ class LLM:
+                    history_list[-1] += f"\nAction: {action_verbose}\n"
 
-        messages += [("user", hist), ("assistant", "Acknowledged.")]
+        obs_act_history += "".join(history_list[-20:])  # Only show up to the last 20 obs act events to avoid cloggage
 
-        # Current observation and action prompt
-        curr_obs = f"Now, the current observation space is {obs_view_full(env_state)} and the changes that have occured since last observation are: {obs_diff(env_state)}"
-        _LOGGER.info("\n\n")
-        _LOGGER.info(f"Observed changes: {obs_diff(env_state)}\n")
+        # Current observation space changes
+        _LOGGER.info(f"{colored('Observed changes', 'yellow')}: {obs_diff(env_state)}\n\n")
 
-        action_prompt = "Please take a suitable action. Action: "
-        messages.append(("user", curr_obs + action_prompt))
+        prompt = REASON_ACTION_SPACE_NODE_SELECT.format(
+            node_names=node_names,
+            service_names=service_names,
+            network_connectivity_desc=network_connectivity_desc(initial_state),
+            initial_obs_view_full=obs_view_full(initial_state),
+            obs_act_history=obs_act_history,
+            current_obs_view_full=obs_view_full(env_state),
+            current_obs_diff=obs_diff(env_state),
+        )
+        messages = [("user", prompt)]
+        prompt = format_llama_prompt(SYSTEM_MSG, messages)
 
-        prompt = format_llama_prompt(system_msg, messages)
-        response = self.client.generate(
-            prompt=prompt,
-            grammar=Grammar(type=GrammarType.Json, value=AgentNodeAction.model_json_schema()),
-            max_new_tokens=300,
+        return prompt
+
+    def _build_action_prompt(
+        self, env_state: EnvironmentState, env_history: list[EnvironmentState], node_name: str, reasoning: str
+    ) -> str:
+        prompt = ""
+
+        # Get the action space description
+        env = env_state.env
+        initial_state = env_history[0]
+        service_names = "'" + ", ".join(env.services_list) + "'"
+
+        # Build the history of actions
+        obs_act_history = "" if env_history else "NO OBSERVATION HISTORY"
+        history_list = []
+        for i, state in enumerate(env_history[1:]):
+            observed_changes = obs_diff(state)
+            action_id = state.action_id
+
+            if observed_changes != "" or action_id:
+                history_list.append(f"\nStep {i}:")
+
+                if observed_changes != "":
+                    history_list[-1] += f"\n{observed_changes}"
+
+                if action_id is not None:
+                    action = NodeAction.from_id(env=env, action_id=action_id)
+                    action_verbose = action.verbose(colored=False)
+                    history_list[-1] += f"\nAction: {action_verbose}\n"
+        obs_act_history += "".join(history_list[-20:])  # Only show up to the last 20 obs act events to avoid cloggage
```

Comment:
there's quite a bit of code repetition here, would be good to split this building into smaller functions and reuse them

Reply:
broken out obs_act_history into util function